contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,816 @@
+is non-destructive.
+1. If the Pod was not created from the target version, the Pod will be deleted 
+and recreated from that version. This form of reconciliation is destructive.
+
+### Specification Updates
+The StatefulSet controller will [snapshot](#snapshot-creation) its target 
+Object state when mutations are made to its `.Spec.Template` or 
+`.Spec.VolumeClaimsTemplate` (Note that the latter is currently immutable).
+
+1. When the StatefulSet controller observes a mutation to a StatefulSet's 
+ `.Spec.Template` it will snapshot its target Object state and compare 
+the snapshot with the version indicated by its `.Status.UpdateVersion`.
+1. If the current state is equivalent to the version indicated by 
+`.Status.UpdateVersion` no update has occurred. 
+1. If the `Status.CurrentVersion` field is empty, then the StatefulSet has no 
+revision history. To initialize its revision history, the StatefulSet controller 
+will set both `.Status.CurrentVersion` and `.Status.UpdateVersion` to the 
+version of the current snapshot. 
+1. If the `.Status.CurrentVersion` is not empty, and if the 
+`.Status.UpdateVersion` is not equal to the version of the current snapshot, 
+the StatefulSet controller will set the `.Status.UpdateVersion` to the version 
+indicated by the current snapshot.
+
+### StatefulSet Revision History
+The StatefulSet controller will use the technique proposed in 
+[Controller History](https://github.com/kubernetes/community/pull/594) to 
+snapshot and version its target Object state.
+
+#### Snapshot Creation
+In order to snapshot a version of its target Object state, it will 
+serialize and store the `.Spec.Template` and `.Spec.VolumesClaimsTemplate` 
+along with the `.Generation` in each snapshot. Each snapshot will be labeled
+with the StatefulSet's `.Selector`.
+
+#### History Reconstruction
+As proposed in 
+[Controller History](https://github.com/kubernetes/community/pull/594), in 
+order to reconstruct the revision history of a StatefulSet, the StatefulSet 
+controller will select all snapshots based on its `Spec.Selector` and sort them 
+by the contained `.Generation`. This will produce an ordered set of 
+revisions to the StatefulSet's target Object state.
+
+#### History Maintenance 
+In order to prevent the revision history of the StatefulSet from exceeding 
+memory or storage limits, the StatefulSet controller will periodically prune 
+its revision history so that no more that `.Spec.RevisionHisotryLimit` non-live 
+versions of target Object state are preserved.
+
+### Update Completion
+The criteria for update completion is as follows.
+
+1. If the StatefulSet's `.Spec.UpdateStrategy.Type` is equal to 
+`OnDeleteStatefulSetStrategyType` then no version tracking is performed. It 
+this case an update can never be in progress.
+1. If the StatefulSet's `.Spec.UpdateStrategy.Type` is equal to 
+`PartitionedStatefulSetStrategyType` updates can not complete. The version 
```

Comment:
One could argue that once your partition is complete then the update is complete.